src/constructs/core/stack.ts

```diff
@@ -140,6 +141,25 @@ export class GuStack extends Stack implements StackStageIdentity, GuMigratingSta
 
       this.tryAddRepositoryTag();
     }
+
+    const { node } = this;
+
+    node.addValidation({
+      validate(): string[] {
+        const NO_ERRORS: string[] = [];
+
+        node
+          .findAll()
+          .filter((construct) => construct instanceof CfnInclude)
+          .map((cfnInclude) => {
+            Annotations.of(cfnInclude).addWarning(
+              `As you're migrating a YAML/JSON template to ${LibraryInfo.NAME}, be sure to check for any stateful resources! Use 'overrideLogicalId' to preserve a resource's logical ID, ensuring it's not removed unintentionally.`
```

Comment:
I wonder if we can be a little clearer here about exactly when this is required? In an ideal world we would scan the list of resources and give the logical ids that we think might be at risk, but I suspect that will be too much effort vs value provided. Perhaps a good compromise would be to provide some examples of stateful resources that people should look to retain (e.g. databases, DNS records, S3 buckets).

On the other hand, we don't want to make people feel nervous unnecessarily! For example, an ALB could be considered a stateful resource but our migration guides actually suggest replacing this (once the new one is serving all traffic).

It's not easy to capture all of this in a short warning message and I think there is also some jargon here that could benefit from further explanation e.g. I suspect that many users will not know what a [logical id](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resources-section-structure.html) is in this context. Consequently, it might be better to link out to a markdown file where we can: explain these nuances, provide a full example of retaining a stateful resource with `overrideLogicalId`, link to relevant AWS docs etc. 

The other advantage of pulling some of this detail into markdown is that we could also link out to it from the migration guide (perhaps from [this section](https://github.com/guardian/cdk/blob/main/docs/migration-guide.md#migrating-resources)), to give users the best possible chance of reading it!

Sorry for the massive comment - what do you think?

Reply:
I think we're both on the same page! I'm in the middle of writing a `stateful-resources.md` file! I'll add that to this PR and then update all the doc changes in #1278 (well https://github.com/guardian/cdk/pull/1278/commits/c7f037ad84f3be52cc463673214ffd6ea6819c82) to link to it, just to keep things DRY.

Reply:
Updated this change, linking to the stateful resources doc from #1291.